core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala

```diff
@@ -395,7 +402,7 @@ final class ShuffleBlockFetcherIterator(
     // then fetch it one more time if it's corrupt, throw FailureFetchResult if the second fetch
     // is also corrupt, so the previous stage could be retried.
     // For local shuffle block, throw FailureFetchResult for the first IOException.
-    while (result == null) {
+    while (!isZombie && result == null) {
```

Comment:
is it possible that `hasNext` returns true and `next` throws `NoSuchElementException`? `isZombie` may get changed by other threads?

Reply:
Yeah that can happen. Right now I think it's 'worse' in that the iterator might be cleaned up and yet next() will keep querying the iterator that's being drained by cleanup().

To really tighten it up I think more or all of `next()` and `cleanup()` would have to be synchronized (?) and I'm not sure what the implications are of that.

We could follow this up with small things like making `hasNext()` synchronized at least, as `isZombie` is marked `GuardedBy("this")`. That still doesn't prevent this from happening but is a little tighter.

@advancedxy what do you think? I think the argument is merely that this fixes the potential issue in 99% of cases, not 100%.

Reply:
> is it possible that hasNext returns true and next throws NoSuchElementException? isZombie may get changed by other threads?

@cloud-fan Yeah, it can happen. But I agree with @srowen. The `isZombie` flag indicates the whole task is finished, there's no point for the consumer of the iterator to be still active. This  changes the semantics of `Iterator` at rare chances,  but I think it is acceptable.

> We could follow this up with small things like making `hasNext()` synchronized at least, as `isZombie` is marked `GuardedBy("this")`. That still doesn't prevent this from happening but is a little tighter.

Maybe. But I would leave it as it's if It's up to me. Like you said, this doesn't prevent the semantics changing but a little tighter.